From b327f6512df4374c57431d295c79ca17b36cc3cf Mon Sep 17 00:00:00 2001
From: yangerkun <yangerkun@huawei.com>
Date: Mon, 15 Nov 2021 19:49:02 +0800
Subject: [PATCH] ext4: avoid recheck extent for EXT4_EX_FORCE_CACHE
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: b327f6512df4374c57431d295c79ca17b36cc3cf


hulk inclusion
category: bugfix
bugzilla: 182952 https://gitee.com/openeuler/kernel/issues/I4DDEL

---------------------------

Buffer with verified means that it has been checked before. No need
verify and call set_buffer_verified again.

Signed-off-by: yangerkun <yangerkun@huawei.com>
Reviewed-by: Jan Kara <jack@suse.cz>
Reviewed-by: Zhang Yi <yi.zhang@huawei.com>

Signed-off-by: Chen Jun <chenjun102@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/ext4/extents.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 618675a41efb..a77e25ca6867 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -544,13 +544,16 @@ __read_extent_tree_block(const char *function, unsigned int line,
 		if (err < 0)
 			goto errout;
 	}
-	if (buffer_verified(bh) && !(flags & EXT4_EX_FORCE_CACHE))
-		return bh;
-	err = __ext4_ext_check(function, line, inode, ext_block_hdr(bh),
-			       depth, pblk, le32_to_cpu(idx->ei_block));
-	if (err)
-		goto errout;
-	set_buffer_verified(bh);
+	if (buffer_verified(bh)) {
+		if (!(flags & EXT4_EX_FORCE_CACHE))
+			return bh;
+	} else {
+		err = __ext4_ext_check(function, line, inode, ext_block_hdr(bh),
+				       depth, pblk, le32_to_cpu(idx->ei_block));
+		if (err)
+			goto errout;
+		set_buffer_verified(bh);
+	}
 	/*
 	 * If this is a leaf block, cache all of its entries
 	 */
-- 
2.26.2

